### DISCUSSION TITLE
Embedded Debugging

### DISCUSSION THREAD

**Discussion by timmyxuereb** [NationalSecurityAgency/ghidra#4563] [Debugger]
Any getting started guide on how to debug an embedded device connected through ghidra?

My issue is with understanding how I can set up a server for gdb or lldb and then connecting to the target. 

The current hex I'm trying to debug is developed on a MIPS32 architecture M4K on the PIC32MX. Debug interfaces available is pickit.

Looked through these but I couldn't understand how to get started
https://gcc.gnu.org/onlinedocs/gcc-11.2.0/gcc/Developer-Options.html
https://wiki.apertus.org/index.php/AXIOM_Remote/Development
https://github.com/sergev/ejtagproxy/blob/master/README.txt
https://www.microchip.com/forums/m346142-p3.aspx
https://forum.sparkfun.com/viewtopic.php?t=28623

**Comment by nsadeveloper789** [comment]
Looks like you've taken a few approaches. For pretty much all non-Windows debug targets, the best way forward is to connect via GDB. If you're unable to get GDB connected to your PIC32MX, then I'm afraid we don't currently have a solution to connect Ghidra to the PIC32MX.

**Comment by d-millar** [comment]
Just as a follow-on, your third reference should allow you to connect with GDB.  Where did you get stuck in that description?

**Reply by timmyxuereb** [reply]
@nsadeveloper789 Yes as @d-millar said I think connecting PIC devices to GDB isn't the issue despite not having managed using the pickit4. (But I think this issue was related to not having the a programmer supported by ejtagproxy)

I was wondering if you have a more broader guide for debugging embedded devices theough ghidra.

**Comment by d-millar** [comment]
@timboxx Sadly, the short answer is no.  The long answer is, as discussed frequently here, that support for embedded devices depends very heavily on the gdbstub built into each specific embedded device.  There is an astonishing amount of variability in underlying implementations and the RSP-protocol packets each gdbstub supports.  For instance, a stub might or might not support breakpoints, specific flavors of breakpoints, might use z/Z for enabling/disabling breakpoints, might use Z/z, might support memory compression in one of several flavors, might support various q commands, including the one that says which packets are supported.  GDB, itself depending on the the version, might accommodate these variations (or might not).  We have tried to standardize on the MI2 interface, but there are a number of commands which have no equivalents in MI2.  In particular, many embedded platforms do not support "info proc mappings" or "maintenance info sections" which we rely on for dynamic-to-static mappings.  If your device does not support these features, there are a few workarounds in Ghidra proper and a few workarounds that involve dummying out those commands (or implementing them).

We are more than happy to try to get you up and running with your particular device, but unfortunately general advice is probably going to be pretty vague and possibly inappropriate for your specific case.  That said, if you can send error messages or other clues, we may be able to give you advice tailored to your situation (particularly if we can recreate it).

**Comment by timmyxuereb** [comment]
Ok thanks for your detailed reply!

I managed to get a pickit3 and updated it's firmware from using the following http://ww1.microchip.com/downloads/en/DeviceDoc/PICkit3%20Programmer%20Application%20v3.10.zip

Requires running the application as an admin and my device turned out to be a bit faulty and needed a shake to finalize updating of an OS.

Now I can get set on using the pickit3 with ejtagproxy route.

I am connecting to gdb but I'm getting the following on gdb: 
![image](https://user-images.githubusercontent.com/47160446/187695475-55a13dd3-2456-45ec-951a-59d4a08093ac.png)
[file has been defined and ejtagproxy running on port 2000]

Connecting by ghidra to gdb instead of console is causing the following error to come up 
![image](https://user-images.githubusercontent.com/47160446/187695851-168590f8-0861-478e-af1a-6f7f162d4834.png)
when using the below settings
![image](https://user-images.githubusercontent.com/47160446/187695951-7a2c683d-c72a-46fe-bfb4-28689003e2bf.png)

Below is an output from: ejtagproxy -d
These are reported after trying to connect to gdb
![image](https://user-images.githubusercontent.com/47160446/187696101-e57f7590-94a6-4984-9620-01b74eb919da.png)

**Reply by timmyxuereb** [reply]
I also have  an Em100Pro-G2 which might allow connection to GDB using openOCD as pickit isn't supported.

**Reply by d-millar** [reply]
Oh, awesome!  Now I think you're close!

You basically did everything right except one little thing, and that, I have to admit, is probably our fault.  "Agent TCP port" is not what you think it is.  That's the port that will be used by Ghidra to talk to the GDB Agent, i.e. the wrapper around the MI2 interface.  Just by way of background, when you choose "IN-VM GU gdb local debugger", Ghidra spins up GDB and executes commands in the same JVM that runs Ghidra.  We do not recommend that generally, even though it's a bit faster, as the debugger may crash and a crash will then take Ghidra (and any work you've done in it) down with it.  If you run "GNU gdb local agent over GADP/TCP" as you did, Ghidra sets up a sort of double client-server:  Ghidra talks to a second Java process, the agent, over a custom protocol (GADP) and the agent talks to the gdbstub over RSP, gdb's native protocol.  The "Agent TCP port" controls the first link, which you probably do not want to mess with.  You want to define the second path, which you do exactly as you did with GDB proper.

In other words, you want to (a) connect to the local agent over GADP using the defaults for eveything except the "GDB launch command", setting that, as you did, to "none-eabi-gdb",  and then (b) in the Interpreter window, execute "target extended-remote localhost:2000".  This _should_ get you connected, and to the point where the "info proc mappings"/"maintenance info sections" stuff will get in your way.  (At that point, you'll probably want to map your static image identically to the dynamic context in the Modules window, but we can give you more info on that once you get to it, and we know exactly where you're getting into trouble.)

**Reply by timmyxuereb** [reply]
If I leave Agent TCP port empty Ghidra will assign a random port. (55589) Not sure I'm understand you though.
![image](https://user-images.githubusercontent.com/47160446/187925808-27984f5f-7347-4bf3-a185-a763964df1f8.png)
Also after whilst it's listening to 55589 the interface doesn't let you back out and use the interpreter
![image](https://user-images.githubusercontent.com/47160446/187926087-2fc74764-602d-408f-976a-70e2a260c8c4.png)

**Reply by timmyxuereb** [reply]
https://github.com/NationalSecurityAgency/ghidra/issues/3549 I saw this issue but its not really solving my issue

**Reply by d-millar** [reply]
@timboxx I'm not sure what you mean when you say "the interface doesn't let you back out and use the interpreter".  Just to be clearer about what I was saying, when Ghidra picks the random port (let's assume 55589), that's a port you having nothing to do with, i.e. it is NOT the port you would use for "target extended-remote".  That port is the port the Ghidra UI is using to communicate with the Ghidra Agent (not the gdbstub in the target).  The port you need for "target extended-remote" is 2000 or whatever the ejtagproxy is exporting.  Are you not seeing an Interpreter windows launched after you connect?  This should be a separate window in the GUI (not the Console or Debug Console) which wraps the gdb CLI.

**Reply by timmyxuereb** [reply]
Yeah that's exactly what I mean, I'm not getting an interpreter window launched and the UI becomes grayed out until I press cancel on this menu/panel/interface
![image](https://user-images.githubusercontent.com/47160446/187935451-2d5f7c3b-8c06-4e11-b2a9-9f93117e88e1.png)

**Reply by d-millar** [reply]
Hmmm.  I don't see anything in the Debug Console indicating why it wouldn't launch.  Is the Agent process running after you connect or has it died?  You should have two Java processes after the launch.  Not sure if you've tried this yet, but running support/ghidraDebug versus ghidraRun will spit a lot more info into a console which might give us more to go on.  And I might try the IN-VM client vs GADP just to see if it makes any of the error messages more obvious.  One thing from your screenshot, you probably do not want 0 for the JDWP port (stick with -1 unless you really want to debug the Agent).

**Reply by d-millar** [reply]
OK, I have a theory as to what might be going wrong, but I need your help to test it out.  If you're game, connect to your device with gdb (no Ghidra), execute the following command, and send us what you get:
     interpreter-exec mi "-thread-info  thread-id"    (where thread-id is the TID of the current thread)

**Reply by d-millar** [reply]
(Sorry about the edit and resend - I forgot brackets aren't a happy thing in these posts.)

**Reply by d-millar** [reply]
interpreter-exec mi "-thread-info TID"

**Reply by timmyxuereb** [reply]
Outputs 

^done, threads=[]

But I'm not managing to connect to the target as I am receiving remote sent bad register. 
I was thinking maybe the memory map i set through ghidra was incorrect? or maybe the hex file should be elf?

**Reply by timmyxuereb** [reply]
By "maybe the hex file should be elf" means that I don't have access to the symbol files. And I'm not sure how I can generate them with the hex file from the disassembled code

**Reply by timmyxuereb** [reply]
https://github.com/aldelaro5/ghidra-ExportDwarfELFSymbols trying this out

**Reply by timmyxuereb** [reply]
Encountered this issue https://github.com/clienthax/Ps3GhidraScripts/issues/19
Solved it by downgrading from JDK 18 to JDK 17. (Removed system variables)

**Comment by d-millar** [comment]
OK, let's back up a little....  Looking over your previous posts, it looks like you haven't been able to get gdb to connect to your target without Ghidra involved.  Is that correct?  I.e. you haven't been able to get past "Remote sent bad register".  I just noticed you're using "arm-none-eabi-gdb".  Am wondering about that - seems like possibly the wrong choice, i.e. isn't that arm-specific?  My background here is a little weak, so maybe you have a better sense of this than I, but am guessing the register error is because the arm variant of gdb is assuming the target is arm, not MIPS32.  Am thinking "gdb-multarch" with the first command being "set architecture mips" might be a better choice.

No point in trying Ghidra until you get gdb working solo, as this will just complicate the debugging process.

**Reply by timmyxuereb** [reply]
Yep you are absolutely right, was about to reply about this; https://github.com/adamrehn/gdb-multiarch-windows/releases Downloaded this release and I'm connecting to the host without issue... Except, now I'm stuck with the issue of symbols. As I can't set breakpoints and looks like the gdb won;t continue .. I'm downloading nasm, and will extract the dissasembled code onto an asm file then i'll generate some symbols. Do you suggest another way?

**Reply by d-millar** [reply]
OK, cool - some progress.  Symbols make things easier long-term, but shouldn't be absolutely necessary for getting things going.  A little more worried about the fact that gdb won't continue.  Can you "stepi"?  Breakpoints may be a very complicated issue.  I don't remember the specifics of MIPS breakpointing, but am guessing you will have to use hardware breakpoints and for now may have to use raw addresses instaed of symbols.

**Reply by timmyxuereb** [reply]
Will get back to you tomorrow. Debugger was getting stuck in address 0xbfc00000 which is the reset handler. Appreciate the support
